Mathlib/GroupTheory/GroupAction/Pointwise.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nathaniel Thomas, Jeremy Avigad, Johannes Hölzl, Mario Carneiro, Anne Baanen,
+  Frédéric Dupuis, Heather Macbeth, Antoine Chambert-Loir
+-/
+
+import Mathlib.Data.Set.Pointwise.SMul
+import Mathlib.GroupTheory.GroupAction.Hom
+
+/-!
+# Pointwise actions of equivariant maps
+
+- `image_smul_setₛₗ` : under a `σ`-equivariant map,
+  one has `h '' (c • s) = (σ c) • h '' s`.
+
+- `preimage_smul_setₛₗ'` is a general version of the equality
+  `h ⁻¹' (σ c • s) = c • h⁻¹' s`.
+  It requires that `c` acts surjectively and `σ c` acts injectively and
+is provided with specific versions:
+  - `preimage_smul_setₛₗ_of_units` when `c` and `σ c` are units
+  - `preimage_smul_setₛₗ` when `σ` belongs to a `MonoidHomClass`and `c` is a unit
+  - `MonoidHom.preimage_smul_setₛₗ` when `σ` is a `MonoidHom` and `c` is a unit
+  - `Group.preimage_smul_setₛₗ` : when the types of `c` and `σ c` are groups.
+
+- `image_smul_set`, `preimage_smul_set` and `Group.preimage_smul_set` are
+the variants when `σ` is the identity.
+
+-/
+
+open Set Pointwise
+
+theorem MulAction.smul_bijective_of_is_unit
+    {M : Type*} [Monoid M] {α : Type*} [MulAction M α] {m : M} (hm : IsUnit m) :
+    Function.Bijective (fun (a : α) ↦ m • a) := by
+  lift m to Mˣ using hm
+  rw [Function.bijective_iff_has_inverse]
+  use fun a ↦ m⁻¹ • a
+  constructor
+  · intro x; simp [← Units.smul_def]
+  · intro x; simp [← Units.smul_def]
+
+variable {R S : Type*} (M M₁ M₂ N : Type*)
+
+variable [Monoid R] [Monoid S] (σ : R → S)
+variable [MulAction R M] [MulAction S N] [MulAction R M₁] [MulAction R M₂]
+variable {F : Type*} (h : F)
+
+-- @[simp] -- In #8386, the `simp_nf` linter complains:
+-- "Left-hand side does not simplify, when using the simp lemma on itself."
+-- For now we will have to manually add `image_smul_setₛₗ _` to the `simp` argument list.
+-- TODO: when lean4#3107 is fixed, mark this as `@[simp]`.
+theorem image_smul_setₛₗ [FunLike F M N] [MulActionSemiHomClass F σ M N]
```

Comment:
Could you please put the next few lemmas in a section with
```lean
variable [FunLike F M N] [MulActionSemiHomClass F σ M N]
variable (c : R) (s : set M) (t : Set N)
```

Reply:
Yes!